### PR TITLE
Feature/external api

### PR DESCRIPTION
**Pull request by torsello**
## Summary
- Consuming external api to get missing books
- If books exists on the api then is persisted on the database

## Feature flags
- status 200: https://i.imgur.com/wNeAVZD.png
- status 201: https://i.imgur.com/FXDNy3W.png
- status 404: https://i.imgur.com/a4uI9eq.png

## Trello Card
- https://trello.com/c/Bfncllvv

### CODE REVIEW

File: src/main/java/com/wolox/training/services/BookService.java [hunk 1]
```diff
@@ -102,4 +132,17 @@ private void adaptBookDtoToBookModel(BookDTO bookDto, Book book) {
...
+    private void adaptBookInfoToBookModel(BookInfoDTO bookInfo, Book book) {
+        book.setIsbn(bookInfo.getIsbn());
+        book.setYear(bookInfo.getPublishDate());
+        book.setAuthor(bookInfo.getAuthors().toString());
+        book.setPublisher(bookInfo.getPublishers().toString());
+        book.setTitle(bookInfo.getTitle());
+        book.setSubtitle(bookInfo.getSubtitle());
+        book.setPages(bookInfo.getNumberOfPages());
+        book.setImage("-");
+        book.setGenre("-");
```
Comment (zeminlu) [inline src/main/java/com/wolox/training/services/BookService.java]: Why? Genre is optional

File: src/main/java/com/wolox/training/services/BookService.java [hunk 2]
```diff
@@ -102,4 +132,17 @@ private void adaptBookDtoToBookModel(BookDTO bookDto, Book book) {
...
+
+    private void adaptBookInfoToBookModel(BookInfoDTO bookInfo, Book book) {
+        book.setIsbn(bookInfo.getIsbn());
+        book.setYear(bookInfo.getPublishDate());
+        book.setAuthor(bookInfo.getAuthors().toString());
+        book.setPublisher(bookInfo.getPublishers().toString());
+        book.setTitle(bookInfo.getTitle());
+        book.setSubtitle(bookInfo.getSubtitle());
+        book.setPages(bookInfo.getNumberOfPages());
+        book.setImage("-");
```
Comment (zeminlu) [inline src/main/java/com/wolox/training/services/BookService.java:144]: Do you need to set this? Aren't you doing it directly in the User class as initialization?

File: src/main/java/com/wolox/training/services/OpenLibraryService.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        String url = new StringBuilder(BASE_URL).append("?bibkeys=ISBN:").append(isbn)
+                .append("&format=json&jscmd=data").toString();
+        BookInfoDTO bookInfo = new BookInfoDTO();
+
+        try{
+            JsonNode bookInfoJsonNode = objectMapper.readTree(restTemplate.getForObject(url, String.class));
+            HashMap<String, Object> mapa = objectMapper.treeToValue(bookInfoJsonNode.get("ISBN:"+isbn), HashMap.class);
+            bookInfo = objectMapper.convertValue(mapa, BookInfoDTO.class);
+
+            if(bookInfo!=null){
```
Comment (zeminlu) [inline src/main/java/com/wolox/training/services/OpenLibraryService.java]: Add missing spacing.

**Review by zeminlu — CHANGES_REQUESTED** [review]
Looks good! Check my comments. I'll recheck later with more time to see if there are any other things to comment.

File: src/main/java/com/wolox/training/services/OpenLibraryService.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    private static final String BASE_URL = "https://openlibrary.org/api/books";
+
+    public Optional<BookInfoDTO> bookInfo(String isbn) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.setVisibility(PropertyAccessor.ALL, Visibility.NONE);
+        objectMapper.setVisibility(PropertyAccessor.FIELD, Visibility.ANY);
+        objectMapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+        RestTemplate restTemplate = new RestTemplate();
+
+        String url = new StringBuilder(BASE_URL).append("?bibkeys=ISBN:").append(isbn)
```
Comment (julianvasquez-wolox) [inline src/main/java/com/wolox/training/services/OpenLibraryService.java]: Maybe you could check the String.format() method for this url. This way you don't need to append so many times and just set the variable into the url string using %s.

File: src/main/java/com/wolox/training/services/OpenLibraryService.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                .append("&format=json&jscmd=data").toString();
+        BookInfoDTO bookInfo = new BookInfoDTO();
+
+        try {
+            JsonNode bookInfoJsonNode = objectMapper
+                    .readTree(restTemplate.getForObject(url, String.class));
+            HashMap<String, Object> mapa = objectMapper
```
Comment (julianvasquez-wolox) [inline src/main/java/com/wolox/training/services/OpenLibraryService.java]: be careful with this spanish-named var😅

**Review by julianvasquez-wolox — COMMENTED** [review]
Seems good! but check the comments 😁 there is something you can improve about the URL

File: src/main/java/com/wolox/training/models/Book.java [hunk 6]
```diff
@@ -34,7 +34,7 @@
     private String author;
 
     @NotNull
-    private String image;
+    private String image = "-";
```
Comment (zeminlu) [inline src/main/java/com/wolox/training/models/Book.java]: I'm sorry I suggested this, I got confused with the DTO that had the variables initialized with this. Your previous approach for `image` was ok, you can revert this.

**Review by zeminlu — CHANGES_REQUESTED** [review]
Looks great, just one comment!

**Review by zeminlu — APPROVED** [review]

File: src/main/java/com/wolox/training/services/OpenLibraryService.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import com.wolox.training.exceptions.MyJsonProcessingException;
+import java.util.HashMap;
+import java.util.Optional;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+@Service
+public class OpenLibraryService {
+
+    private static final String BASE_URL = "https://openlibrary.org/api/books?bibkeys=%s&format=json&jscmd=data";
```
Comment (julianvasquez-wolox) [inline src/main/java/com/wolox/training/services/OpenLibraryService.java]: Look great, let's do something extra... read this value from the application properties. It's something very used, you can read the value using @Value("${path.to.value}") 😁

**Review by julianvasquez-wolox — APPROVED** [review]

**Pull request merged**